### DISCUSSION TITLE
debugging a feature

### DISCUSSION THREAD

**Discussion by daggs1** [endless-sky/endless-sky#8319] [Balance, Mechanics, and Interactions]
Greetings,

I'm testing a feature which seems to end in crash be cause one of the ships in the list get emptied for some reason.
any hints on debugging this issue? does the game have other debugging tools beside the logs?

**Comment by quyykk** [comment]
How to you build the game? And on what OS?

**Reply by daggs1** [reply]
linux, scons

**Reply by RisingLeaf** [reply]
gdb would be my choice.
a bit of explanation on what feature would help though

**Reply by daggs1** [reply]
already using it however I'm not sure what to make of it.
the feature is https://github.com/endless-sky/endless-sky/pull/7906 + fixes to that PR + same logic for ships and shipyards.
code can be seen at https://github.com/daggs1/endless-sky/tree/full_features

**Reply by quyykk** [reply]
What's your distro? And version?

You might find it easier to download vscode and installation couple of extensions (namely the C++ and CMake extensions from Microsoft). That gets you a complete development environment with debugging support.

**Reply by daggs1** [reply]
using Gentoo, kdevelop as IDE
logs show that at one occasion, a entry in visitedPlanets has a null planet

**Reply by quyykk** [reply]
kdevelop seems to have an integrated debugger, you could try that. It will give you a stack trace so you can pinpoint exactly where the crash happens.

**Reply by RisingLeaf** [reply]
> using Gentoo, kdevelop as IDE logs show that at one occasion, a entry in visitedPlanets has a null planet

I use kdevelop on linux and the integrated debugger is really good.
<del>also, gentoo? I think i found your problem</del>

**Reply by daggs1** [reply]
> > using Gentoo, kdevelop as IDE logs show that at one occasion, a entry in visitedPlanets has a null planet
> 
> I use kdevelop on linux and the integrated debugger is really good. ~also, gentoo? I think i found your problem~

I need to understand at what stage the planet gets nulled, unless I'm wrong, that cannot be done in gdb (that said, my day job is C in non standard envs so gdb is not something I use)

as for the gentoo remark, gentoo > whatever you are using :p.
on another matter, can you please review https://github.com/endless-sky/endless-sky/pull/7906 which awaits your review for more than two weeks please?

**Reply by daggs1** [reply]
placing a watch on the map seems to cause the game to freeze on resume
I've looked at the code. the issue is when saving the visitedPlanets, the sort function in WriteSorted encounters a null Planet.
I have changes in visitedPlanets however, usage of the key (Planet) is always due as before the change, e.g. passed by ref and casted into ptr.
I don't see how it is possible to get NULL Planet. I'm leaning into a possible compiler bug?

thoughts?
